service/history/workflow/mutable_state_impl.go

```diff
@@ -5107,6 +5109,22 @@ func (ms *MutableStateImpl) closeTransactionHandleActivityUserTimerTasks(
 	}
 }
 
+func (ms *MutableStateImpl) closeTransactionCollapseUpsertVisibilityTasks() {
+	// check if we have >= 2 tasks that are all upsert visibility with the same version
+	visTasks := ms.InsertTasks[tasks.CategoryVisibility]
+	if len(visTasks) < 2 {
+		return
+	}
+	for _, task := range visTasks {
+		if task.GetType() != enumsspb.TASK_TYPE_VISIBILITY_UPSERT_EXECUTION ||
+			task.GetVersion() != visTasks[0].GetVersion() {
```

Comment:
I remember version is not actually used when processing this task. @alexshtin Can you confirm?

Reply:
It's not. There's a comment somewhere else that says so. And I can't imagine any way we could end up with two versions here. But I wanted to be conservative

Reply:
ideally I think we should compare all fields in the UpsertVisibilityTask definition in case new fields got added in the future and tasks accidentally got collapsed here.

Reply:
Using the Task interface or UpsertVisibilityTask directly? I can try that...